### PR TITLE
Cpp extension

### PR DESCRIPTION
**Pull request by goldsborough**
Moves the tutorial for using C++ extensions here, from https://github.com/pytorch/pytorch/pull/5541.

Reviews still welcome. @apaszke @zdevito @soumith 

I've also removed the C extensions tutorial, in order to promote the C++ extensions, which are a superset of functionality and easier to use. Let me know if this is not OK.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -105,3 +105,4 @@ ENV/
 
 # Rope project settings
 .ropeproject
+intermediate_source/
```
Comment (chsasank) [inline .gitignore]: Why is this added to gitignore?

Comment (goldsborough) [reply .gitignore]: I can't remember right now, but it might have been temporary output from the docs build that I wanted to not commit.

File: advanced_source/cpp_extension.rst [hunk 2]
```diff
@@ -0,0 +1,1033 @@
...
+
+More performance increases!
+
+Conclusion
+----------
+
+You should now be equipped with a good overview of PyTorch's C++ extension
+mechanism as well as a motivation for using them. You can find the code examples
+displayed in this note `here
+<https://github.com/goldsbborough/pytorch-cpp-extension>`_. If you have
```
Comment (t-vi) [inline advanced_source/cpp_extension.rst]: Is this intentionally not the link to https://github.com/pytorch/extension-cpp/ ?

**Comment by RoyiAvital** [comment]
Could we leave option for C only extensions?
Some would prefer sticking with C.

File: advanced_source/cpp_extension.rst (hunk 2 above)
Comment (goldsborough) [reply advanced_source/cpp_extension.rst]: Nope, this needs to be fixed.

**Comment by goldsborough** [comment]
@RoyiAvital Not really, for one we're trying to phase out TH* things wherever we can, and for the other getting rid of `cffi` would be a blessing and make everyone's life much nicer. No one forces you to use classes with C++ extensions :)

File: advanced_source/cpp_extension.rst [hunk 3]
```diff
@@ -0,0 +1,1033 @@
...
+
+  class LLTM(torch.nn.Module):
+      def __init__(self, input_features, state_size):
+          super(LLTM, self).__init__()
+          self.input_features = input_features
+          self.state_size = state_size
+          # 3 * state_size for input gate, output gate and candidate cell gate.
+          # input_features + state_size because we will multiply with [input, h].
+          self.weights = torch.nn.Parameter(
+              torch.Tensor(3 * state_size, input_features + state_size))
```
Comment (SsnL) [inline advanced_source/cpp_extension.rst]: Can we use `torch.empty` instead of `torch.Tensor`?

File: advanced_source/cpp_extension.rst [hunk 4]
```diff
@@ -0,0 +1,1033 @@
...
+  import torch
+
+  assert torch.cuda.is_available()
+
+  batch_size = 16
+  input_features = 32
+  state_size = 128
+
+  # Note the .cuda() calls here
+  X = torch.randn(batch_size, input_features).cuda()
```
Comment (SsnL) [inline advanced_source/cpp_extension.rst]: Use the new `device=` option please.

File: advanced_source/cpp_extension.rst [hunk 5]
```diff
@@ -0,0 +1,1033 @@
...
+
+  forward = 0
+  backward = 0
+  for _ in range(100000):
+      start = time.time()
+      new_h, new_C = rnn(X, (h, C))
+      forward += time.time() - start
+
+      start = time.time()
+      (new_h.sum() + new_C.sum()).backward()
```
Comment (SsnL) [inline advanced_source/cpp_extension.rst]: Don't we need synchronization to get correct timings?

Comment (goldsborough) [reply advanced_source/cpp_extension.rst]: what?
